scanner/src/main/kotlin/experimental/WorkingTreeCache.kt

```diff
@@ -64,13 +64,15 @@ class DefaultWorkingTreeCache : WorkingTreeCache {
         return getWorkingTreeMutex(vcsInfo).withLock { block(vcs, getWorkingTree(vcsInfo, vcs)) }
     }
 
-    private fun getKey(vcsInfo: VcsInfo) = "${vcsInfo.type}|${vcsInfo.url}|${vcsInfo.path}"
+    private fun getMutexKey(vcsInfo: VcsInfo) = "${vcsInfo.type}|${vcsInfo.url}"
+
+    private fun getWorkingTreeKey(vcsInfo: VcsInfo) = "${vcsInfo.type}|${vcsInfo.url}|${vcsInfo.path}"
 
     private suspend fun getWorkingTreeMutex(vcsInfo: VcsInfo) =
         mutex.withLock {
             check(!terminated) { "The cache was already shut down." }
 
-            workingTreeMutexes.getOrPut(getKey(vcsInfo)) { Mutex() }
+            workingTreeMutexes.getOrPut(getMutexKey(vcsInfo)) { Mutex() }
```

Comment:
I don't understand the existing approach, in particular why multiple working trees can exist for same (type, url) but with different path (or also revision). Mind giving me a hint?

(I wonder whether switching to max 1 working tree per (type, url) would be better)

Reply:
The historic reason was that the sparse checkout configuration is written to a file, so I wanted to avoid issues with re-using the working tree with and without sparse checkout. This is currently a draft because I just wanted to quickly see if this would fix our issue. I will alter replace it with another change that does indeed remove sparse checkouts. This is something I had on my list for some time because the current implementation clones the same repository once for each project found in a different path.

Reply:
It's updated now.